File: nodejs/day11/index2.js
const { readFile } = require("../lib");

const data = readFile("./input").split(",");

const mapDir = dir => {
  switch (dir) {
    case "n":
      return { x: 1, y: 0, z: -1 };
    case "ne":
      return { x: 1, y: -1, z: 0 };
    case "nw":
      return { x: 0, y: 1, z: -1 };
    case "sw":
      return { x: -1, y: 1, z: 0 };
    case "s":
      return { x: -1, y: 0, z: 1 };
    case "se":
      return { x: 0, y: -1, z: 1 };
  }
};

let c = { x: 0, y: 0, z: 0 };
let max = 0;

data.map(mapDir).forEach(({ x, y, z }) => {
  c = {
    x: x + c.x,
    y: y + c.y,
    z: z + c.z
  };
  const d = (Math.abs(c.x) + Math.abs(c.y) + Math.abs(c.z)) / 2;
  max = Math.max(max, d);
});

console.log(max);

File: clojure/day11/index2.clj
(require '[clojure.string :as s])

(def parse-int #(Integer/parseInt %))
(defn split [p s] (s/split s p))

(defn map-dir [dir]
  (case dir
    "n"  { :x 1  :y 0  :z -1 }
    "ne" { :x 1  :y -1 :z 0 }
    "nw" { :x 0  :y 1  :z -1 }
    "sw" { :x -1 :y 1  :z 0 }
    "s"  { :x -1 :y 0  :z 1 }
    "se" { :x 0  :y -1 :z 1 }))

(def data (->>
  "./input"
  slurp
  s/trim-newline
  (split #",")
  (map map-dir)))

(defn add-coords [{x1 :x y1 :y z1 :z} {x2 :x y2 :y z2 :z}]
  {:x (+ x1 x2) :y (+ y1 y2) :z (+ z1 z2)})

(def res (loop [
  c {:x 0 :y 0 :z 0}
  m 0
  rem-data data
] (if (== 0 (count rem-data)) m (let [
  c' (first rem-data)
  rem-data' (rest rem-data)
  c'' (add-coords c c')
  {:keys [x y z]} c''
  d (/ (+ (Math/abs x) (Math/abs y) (Math/abs z)) 2)
  m' (max m d)
](recur c'' m' rem-data')))))

(println res)